### ISSUE TITLE
code for equipment / gear types needs refactoring

### ISSUE THREAD

**Issue by Alys** [HabitRPG/habitica#5441]
There is code that lists every kind of equipment and handles each kind more or less identically. This is unpleasant when a new equipment type is added (and in fact some of the new equipment types probably aren't represented in all appropriate places in the code).

Here is one example:
https://github.com/HabitRPG/habitrpg/blob/fc9f511542ea39ec6e099ee3f491ee735c400e00/website/public/js/controllers/rootCtrl.js#L217
and there's another one a couple of lines down from that. There's probably others elsewhere.

It would be really nice if all such code could be refactored so that there's ONE array/object/whatever that lists all the gear types, with config that specifies their env.t string and any special handling that each one might need. Any bits of actual code could then use that array in a neat loop or something. 

Pretty please.

Finding all such code would be a good job for a new contributor. Refactoring it might be too, or might be suitable for a mid-level dev. I'm advertising this in the Aspiring Coders guild.

References to show why this would be good:
https://github.com/HabitRPG/habitrpg/commit/fc9f511542ea39ec6e099ee3f491ee735c400e00
https://github.com/HabitRPG/habitrpg/commit/c6256d7fa109c3e2c74e83334106112a03547687

**Comment by cavejay** [comment]
This is something that I could look into.
I already have 1 contrib tier, so if someone wanted to tag-team and find the code that'd be great.
If not, when I work on this over the next couple of days I'll take a look around.

I'm still not incredibly familiar with the HabitRPG codebase, so if I were to create an object with all the equipment types, where would be the best place to put it?

**Comment by Alys** [comment]
Thank you!

The equipment pieces are defined at https://github.com/HabitRPG/habitrpg/blob/develop/common/script/content.coffee#L58
so probably just above there would work. We can always move it later if it turns out there's a better place for it.

**Comment by cthecheese** [comment]
I wanted to see where @no1mzb was at with this project. I'd have no qualms with assisting in the refactoring.

**Comment by cavejay** [comment]
Starting and creating a pull request today hopefully :)

**Comment by no1mzb** [comment]
@Alys why is there only a .toLowerCase() on only the Saddle check?

https://github.com/HabitRPG/habitrpg/blob/fc9f511542ea39ec6e099ee3f491ee735c400e00/website/public/js/controllers/rootCtrl.js#L217

**Comment by cavejay** [comment]
A problem I'm running into with creating the object with all the equipment in content.coffee is that window doesn't exist in that scope, so calls for 'window.env.t('weapon') fail.

What would be the best way to resolve this? Would it just be to move the object into a file that includes window?

**Comment by Alys** [comment]
You might not need to actually make a call to env.t from content.coffee, but instead just put the item's locale key into the object, and then env.t can be called from whichever jade files use the object. The main point of this change is to let the jade files have loop through all equipment types, rather than having duplicated code for each equipment type individually, so having a call to env.t within the loop will be fine.

BTW, if you'd like feedback on your code at any point, you could push it to your github fork and let us look at it. It doesn't need to be completed before we see it. Don't feel that you have to do that though.

**Comment by cavejay** [comment]
The attached link is to my branch and shows the object just above the code that would be used to search/produce the correct string. https://github.com/HabitRPG/habitrpg/compare/develop...cavejay:equipment_refactor

Would you prefer a proper if statement be used, or is the ternary okay?

Also, in your original issue you said "and any special handling that each one might need". In this case that's only the foodSaddleText env.t key which apparently needs to be converted to lower case before being shown. To cope with that though wouldn't some sort of check still need to be done outside of the object, which is what we're trying to avoid?

Being mostly new to js I was thinking that it should be possible to add a function to equipmentList that could be passed window or window.env and return the appropriate string. It wouldn't be more pollution of the name spaaaaaace, as it would also be part of the equipmentList object. Is that something that you would recommend? Or is that becoming too much of a 'hacky' solution?

**Comment by hairlessbear** [comment]
@cavejay I don't think that's quite what @Alys was looking for. As she said [here](https://github.com/HabitRPG/habitrpg/issues/5441#issuecomment-113979156):

> The equipment pieces are defined at https://github.com/HabitRPG/habitrpg/blob/develop/common/script/content.coffee#L58
> so probably just above there would work. We can always move it later if it turns out there's a better place for it.

If I understand correctly, what Alys is looking for is code to be added (probably in the above-mentioned `content.coffee`, but possibly elsewhere) that makes an array/object/whatever of all equipment types, which is then made available to all other files that may need it. In what you've done so far, you created an object that has all of the equipment types, but it's being created in `rootCtrl.js`, and thus is only accessible by the functions in `rootCtrl.js`. This means that the same object would have to be created in every file that needs it, which is what is currently done (but with a string), and exactly what needs to be changed. Does that make sense?

**Comment by cavejay** [comment]
@hairlessbear yes, I realise this. I was showing the proposed object format and code to access/read/use it. There's no pull request yet because I know that's not where the object isn't going to stay. It's in the format it is because (with the object just above the code) I can easily check if the system will work.

I have a good grasp of code/scoping :) and I am well aware that the code I linked to isn't the final solution to  this issue.

**Comment by Alys** [comment]
@cavejay For special handling, I'd like the object to be able to handle anything special that comes up in future, as well as what we have now, so it needs to be easily expandable. That's best achieved with key-value pairs within the object. For example, you could have an attribute in the object for specifying that a name must be lower-cased, as well as one for specifying the locale key. Also, since you're including things that aren't equipment (probably a good idea), have a true/false key for isEquipment so that we can do special handling for all equipment if we need to.

**Comment by cavejay** [comment]
@Alys still no pull request but this is official draft before that happens.
The refactoring currently looks like this: https://github.com/HabitRPG/habitrpg/compare/develop...cavejay:equipment_refactor
The object is stored in content.coffee and in rootctrl.js I've gone over 3 different ways that the object could be used. I personally prefer the first implementation as it means very little code for each time the string is fetched. I don't know how the standard for doing things like passing around the window object though, so while I might like it the best I'm totally happy to that idea shot down.

Apologies @hairlessbear if my response to your comment came across negatively, I was somewhat frustrated because you didn't have any input to the mulitple questions I had asked, while telling me information I already knew. I appreciate that you took the time to write a potentially important and helpful comment! :)

tl;dr please look at rootctrl.js (here: https://github.com/HabitRPG/habitrpg/compare/develop...cavejay:equipment_refactor) and tell me which one you like the best (and any helpful thoughts/suggestions)

In future should I just make a pull request and link it, regardless of how pull-ready it is? I feel that links the code to this issue much nicer than me continually posting links.

**Comment by hairlessbear** [comment]
@cavejay No worries; it happens :)

I can answer your question about the pull request though: yes, you should create a pull request, even if it's incomplete (it even says so [on the wiki](http://habitrpg.wikia.com/wiki/Guidance_for_Blacksmiths#Pull_Request)). You're right that it's easier to go over the code and what needs to be changed in that view. It's also helpful because it keeps all comments about what/how to implement in one place, making the conversation easier to follow (even if the conversation gets [reeeeeeeeeeeeally long](https://github.com/HabitRPG/habitrpg/pull/5298)).

**Comment by cavejay** [comment]
There pull request is made so code is easier to view. 

@hairlessbear do you have any opinion at all about the varying pieces of code in rootctrl ? Our are you specifically avoiding giving one?

**Comment by hairlessbear** [comment]
Oh I definitely prefer the first option. It's simple and clear.

**Comment by Alys** [comment]
I also like the first option (`var string = (type in equipmentList) ? equipmentList.get(type, window) : type;`). I think it's very neat.

For `needsLowerCase`, I'm wondering if it's worth leaving out the false values for clarity. The code shouldn't need them (or am I wrong?)  and since it's a once-off exception, I think it's clearer what's happening if it's not specified at all for the other items. Opposing opinions are welcome. :)

`equipmentList` should be renamed since this is no longer just for equipment; sorry, I should have mentioned that earlier. `itemList` maybe?

I like how this is going! It will be a lovely improvement to that code.

**Comment by Alys** [comment]
Discussion is now occurring in https://github.com/HabitRPG/habitrpg/issues/5503 rather than here.

**Comment by cavejay** [comment]
Progress was made towards this issue and merged.
Should the Issue be closed?

**Comment by Alys** [comment]
Yes, thanks!
